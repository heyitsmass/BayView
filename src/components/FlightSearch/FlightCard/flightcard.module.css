.flightCard {
    @apply w-full max-w-lg p-6 bg-white border border-zinc-200 rounded-lg shadow dark:bg-zinc-800 dark:border-zinc-700;
}
