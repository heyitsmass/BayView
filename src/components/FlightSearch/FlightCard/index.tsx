import { Flight } from "@/types/ItineraryManager";
import { faCalendarDays, faLocationArrow, faArrowRight, faUser, faChair } from "@fortawesome/free-solid-svg-icons";
import styles from "./flightcard.module.css";
import Input from "@/components/FlightSearch/Input";
import InputPair from "@/components/FlightSearch/Input/InputPair";
import SubmitButton from "@/components/FlightSearch/SubmitButton";

type DataRequest = {
    origin: string;
    destination: string;
    departure_date: string;
    return_date: string;
    number_of_passengers: number;
};

export default async function Page() {
	async function submitSearch(formData: FormData) {
		"use server";

		// mutate data request
		const dataReq: DataRequest = {
			origin: formData.get("departure") as string,
			destination: formData.get("destination") as string,
			departure_date: formData.get("departure") as string,
			return_date: formData.get("return-date") as string,
			number_of_passengers: Number(formData.get("number_of_passengers")),
		};

		// run search
		// const flightResult = await Flight.find_cheapest_flight(
		// 	dataReq.origin,
		// 	dataReq.destination,
		// 	dataReq.departure_date,
		// 	dataReq.return_date,
		// 	dataReq.number_of_passengers
		// );
	}

	return (
		<div>
			<div className={styles.flightCard}>
				<h5 className="text-2xl pb-4 font-medium text-zinc-900 dark:text-white">Flights</h5>
				<form className="space-y-4" action={submitSearch}>
					<InputPair inlineicon={faArrowRight}>
						<Input label="Departing From" name="departure" inlineicon={{ icon: faLocationArrow }} placeholder="e.g., New York" required />
						<Input label="Arriving To" name="destination" inlineicon={{ icon: faLocationArrow }} placeholder="e.g., Los Angeles" required />
					</InputPair>
					<InputPair inlineicon={faArrowRight}>
						<Input
							label="Departure Date"
							name="departure-date"
							inlineicon={{ icon: faCalendarDays }}
							placeholder="Date"
							required
						/>
						<Input label="Return Date" name="return-date" inlineicon={{ icon: faCalendarDays }} placeholder="Date" required />
					</InputPair>
					<InputPair>
						<Input
							label="Number Passengers"
							name="number_of_passengers"
							inlineicon={{ icon: faUser }}
							type="number"
							placeholder="Number Passengers"
							required
						/>
						<Input label="Cabin Preference" name="cabin_preference" inlineicon={{ icon: faChair }} placeholder="Cabin Preference" />
					</InputPair>
					<SubmitButton />
				</form>
			</div>
		</div>
	);
}
