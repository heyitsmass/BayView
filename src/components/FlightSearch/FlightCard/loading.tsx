 export default function loading() {
    return <p>loading...</p>
 }