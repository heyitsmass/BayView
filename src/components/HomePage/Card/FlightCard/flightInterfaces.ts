import { Offer, OfferRequest, OfferRequestSlice, Places } from '@duffel/api/types';

// Define actions the dispatch function will accept
export interface FlightActionInterface {
	type: string;
	payload?: any;
}

// Augment the existing interface to include total duration
declare module '@duffel/api/types' {
	export interface Offer {
		total_duration: string;
	}
}

export interface FlightStateInterface {
	step: number;
	searchState: {
		data: {} | null;
		formValues: {
			departure_date: string;
			return_date: string;
			origin: string;
			destination: string;
			class: string;
			number_of_passengers: number;
		};
	};
	selectState: {
		data: {
			offers: Offer[];
			slices: OfferRequestSlice[];
			id: string;
		} | null;
		formValues: {
			results_sorting: string;
		};
		selectedOffer: {
			offer: Offer;
			id: string;
		} | null;
		returnFlight: {
			data: {
				offers: Offer[];
				slices: OfferRequestSlice[];
				id: string;
			} | null;
			formValues: {
				results_sorting: string;
			};
			selectedOffer: {
				offer: Offer;
				id: string;
			} | null;
		};
	};
	bookState: {};
	confirmedState: {};
	places: Places[];
	roundTrip: boolean;
	cardTitle: string;
	cardSubtitle: string;
	contentPending: boolean;
	isExiting: boolean;
}
