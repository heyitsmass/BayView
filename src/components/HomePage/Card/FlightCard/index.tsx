import {
  faCalendarDays,
  faLocationArrow,
  faArrowRight,
  faUser,
  faChair,
} from "@fortawesome/free-solid-svg-icons";
import Input from "@/components/Input";
import InputPair from "@/components/Input/InputPair";
import Button from "@/components/Button";
import Card from "@/components/HomePage/Card";

export default function FlightCard() {
  return (
    <Card title="Flights" subtitle="search, book, and add to itinerary">
      <form>
        <InputPair icon={faArrowRight}>
          <Input
            label="Departing From"
            name="departure"
            icon={{ icon: faLocationArrow }}
            placeholder="e.g.  New York"
            required
          />
          <Input
            label="Arriving To"
            name="destination"
            icon={{ icon: faLocationArrow }}
            placeholder="e.g.  Los Angeles"
            required
          />
        </InputPair>
        <InputPair icon={faArrowRight}>
          <Input
            label="Departure Date"
            name="departure-date"
            icon={{ icon: faCalendarDays }}
            placeholder="Date"
            required
          />
          <Input
            label="Return Date"
            name="return-date"
            icon={{ icon: faCalendarDays }}
            placeholder="Date"
            required
          />
        </InputPair>
        <InputPair>
          <Input
            label="Number Passengers"
            name="number_of_passengers"
            icon={{ icon: faUser }}
            type="number"
            placeholder="Passengers"
            required
          />
          <Input
            name="cabin_preference"
            icon={{ icon: faChair }}
            placeholder="Cabin Preference"
          />
        </InputPair>
        <Button variant="secondary" className="mt-4">
          Find Flights
        </Button>
      </form>
    </Card>
  );
}
