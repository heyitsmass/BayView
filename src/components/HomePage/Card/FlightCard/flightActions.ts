'use server';
import { Duffel, DuffelError } from '@duffel/api';
import { FlightStateInterface } from './flightInterfaces';
import { CabinClass } from '@duffel/api/types';

// Note: Proper paging functionality is not implemented, client sorting is preferred.
// The current design fetches all offers and then acts upon displayed data with client filtering and sorting.

const flightAPI = new Duffel({
	token: process.env.DUFFEL_API_TOKEN || '',
	debug: { verbose: false },
});

async function validateNameandIATACode(name) {
	try {
		const response = await flightAPI.suggestions.list({
			name: name,
		});
		return response.data[0];
	} catch (error: unknown) {
		if (error instanceof DuffelError) {
			throw error;
		}
		throw new Error('Something went wrong' + error); // Throwing a generic error if it's not a DuffelError
	}
}

export async function getSingleOfferById(id: string) {
	try {
		const response = await flightAPI.offers.get(id);

		return { offer: response.data };
	} catch (error: unknown) {
		if (error instanceof DuffelError) {
			throw error;
		}
		throw new Error('Something went wrong' + error); // Throwing a generic error if it's not a DuffelError
	}
}

// Default action for creating offer request
export async function flightSearchAction(formState: FlightStateInterface['searchState']['formValues']) {
	// populate / mutate data request
	let dataReq = {
		origin: await validateNameandIATACode(formState.origin),
		destination: await validateNameandIATACode(formState.destination),
		departure_date: new Date(formState.departure_date).toISOString(),
		return_date: '', // Field empty if one-way
		number_of_passengers: formState.number_of_passengers,
		class: formState.class as CabinClass,
	};

	// Meta pertaining to offer requests
	const meta = [
		{
			limit: 10,
		},
	];

	let slices;
	let offerRequest;

	// If a return date is specified create a partial offer request for the selected data
	// indicating a multi-step route. Otherwise, create a full-offer request for the one way flight with an empty return date
	if (formState.return_date.length === 0) {
		slices = [
			{
				origin: dataReq.origin.iata_code as string,
				destination: dataReq.destination.iata_code as string,
				departure_date: dataReq.departure_date,
				return_date: '',
			},
		];

		// Create a full offer request
		offerRequest = flightAPI.offerRequests.create({
			meta,
			slices,
			passengers: [{ type: 'adult' }], // Assuming all passengers are 21 for now
			return_offers: true,
			cabin_class: dataReq.class,
		});
	} else {
		// Convert the return field
		dataReq.return_date = new Date(formState.return_date).toISOString();

		// Slices for both departing and returning
		slices = [
			{
				origin: dataReq.origin.iata_code as string,
				destination: dataReq.destination.iata_code as string,
				departure_date: dataReq.departure_date,
			},
			{
				origin: dataReq.destination.iata_code as string,
				destination: dataReq.origin.iata_code as string,
				departure_date: dataReq.return_date,
			},
		];

		// Create a partial offer request
		offerRequest = flightAPI.partialOfferRequests.create({
			meta,
			slices,
			passengers: [{ type: 'adult' }], // Assuming all passengers are 21 for now
			return_offers: true,
			cabin_class: dataReq.class,
		});
	}

	try {
		const response = await offerRequest;

		return { data: response.data, formValues: dataReq};
	} catch (error: unknown) {
		if (error instanceof DuffelError) {
			throw error;
		}
		throw new Error('Something went wrong' + error); // Throwing a generic error if it's not a DuffelError
	}
}

export async function getNextSliceFromPartialOfferRequestIDAction(requestId: string, offerId: string) {
	try {
		const response = await flightAPI.partialOfferRequests.get(requestId,{
			'selected_partial_offer[]': [offerId]
		  });
		return { data: response };
	} catch (error: unknown) {
		if (error instanceof DuffelError) {
			throw error;
		}
		throw new Error('Something went wrong' + error); // Throwing a generic error if it's not a DuffelError
	}
}
