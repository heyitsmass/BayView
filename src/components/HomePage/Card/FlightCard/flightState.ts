'use client';
import { Dispatch, createContext, useContext, useReducer } from 'react';
import { FlightActionInterface, FlightStateInterface } from './flightInterfaces';
import { stepDetails } from '.';

// Helper function for selective deep merging
function deepMerge(target, source) {
	for (const key of Object.keys(source)) {
		// If source[key] is an array or a primitive, replace it entirely
		if (Array.isArray(source[key]) || !(source[key] instanceof Object)) {
			target[key] = source[key];
		} else {
			// If the key doesn't exist in the target, or isn't an object, replace it
			if (!(key in target) || !(target[key] instanceof Object)) {
				target[key] = {};
			}
			// Recursive call for object
			deepMerge(target[key], source[key]);
		}
	}
	return target;
}

// Create the context with the correct type for dispatch
export const FlightContext = createContext<{
	flightState: FlightStateInterface;
	dispatch: Dispatch<FlightActionInterface>;
}>({
	flightState: {
		step: 1,
		searchState: {
			data: null,
			formValues: {
				departure_date: '',
				return_date: '',
				origin: '',
				destination: '',
				number_of_passengers: 1,
				class: 'economy',
			},
		},
		selectState: {
			data: null,
			formValues: {
				results_sorting: 'total_amount',
			},
			selectedOffer: null,
			returnFlight: {
				data: null,
				formValues: {
					results_sorting: 'total_amount',
				},
				selectedOffer: null,
			},
		},
		bookState: {},
		confirmedState: {},
		places: [],
		roundTrip: true,
		cardTitle: '',
		cardSubtitle: '',
		contentPending: false,
		isExiting: false,
	},
	dispatch: () => {},
});

export const FlightReducer = (state, action) => {
	switch (action.type) {
		case 'SET_STEP':
			return {
				...state,
				step: action.payload,
				cardTitle: stepDetails[action.payload].title,
				cardSubtitle: stepDetails[action.payload].subtitle,
				isExiting: true
			};

		case 'UPDATE_CARD_TITLES':
			return {
				...state,
				cardTitle: action.payload.title,
				cardSubtitle: action.payload.subtitle,
			};

		case 'UPDATE_FORM_DATA':
			return {
				...state,
				formData: { ...state.formData, ...action.payload },
			};

		case 'UPDATE_FLIGHT_STATE':
			return deepMerge({ ...state }, action.payload);

		case 'UPDATE_SELECTED_DEPARTING_OFFER':
			return {
				...state,
				selectState: {
					...state.selectState,
					selectedOffer: action.payload, // action.payload should be the new selectedOffer object
				},
			};

		case 'UPDATE_SELECTED_RETURNING_OFFER':
			return {
				...state,
				selectState: {
					...state.selectState,
					returnFlight: {
						...state.selectState.returnFlight,
						selectedOffer: action.payload, // action.payload should be the new selectedOffer object for the return flight
					},
				},
			};

		case 'SET_EXITING':
			return {
				...state,
				isExiting: action.payload
			};

		// Handle updates for searchState, bookState, confirmedState, selectState
		case 'UPDATE_SEARCH_STATE':
			return {
				...state,
				searchState: { ...state.searchState, ...action.payload },
			};

		case 'UPDATE_BOOK_STATE':
			return {
				...state,
				bookState: { ...state.bookState, ...action.payload },
			};

		case 'UPDATE_CONFIRMED_STATE':
			return {
				...state,
				confirmedState: { ...state.confirmedState, ...action.payload },
			};

		case 'UPDATE_SELECT_STATE':
			return {
				...state,
				selectState: { ...state.selectState, ...action.payload },
			};

		default:
			return state;
	}
};

export const useFlightCardState = () => {
	const initialContext = useContext(FlightContext).flightState;
	const initialState = {
		...initialContext,
		step: 1,
		cardTitle: stepDetails[1].title,
		cardSubtitle: stepDetails[1].subtitle,
	};

	const [flightState, dispatch] = useReducer(FlightReducer, initialState);
	return { flightState, dispatch };
};
