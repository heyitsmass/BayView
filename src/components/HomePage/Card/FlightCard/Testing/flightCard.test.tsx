import { describe, test, expect, beforeEach, vi } from 'vitest';
import { render } from '@testing-library/react';
import FlightCard from '../../index';
import { flightSearchAction, getNextSliceFromPartialOfferRequestIDAction, getSingleOfferById } from '../flightActions';

describe('FlightCard', () => {
	test('FlightCard component snapshot', () => {
		const { container } = render(<FlightCard title="Test" />);
		expect(container).toMatchSnapshot();
	});

	describe('flightSearchAction', () => {
		test('one-way flight search branch is triggered when return_date is empty', async () => {
			// Test data with empty return_date
			const formState = {
				origin: 'LAX',
				destination: 'JFK',
				departure_date: '2023-01-01',
				return_date: '',
				number_of_passengers: 1,
				class: 'economy',
			};

			// Call the function
			let errorThrown = false;
			try {
				await flightSearchAction(formState);
			} catch (error) {
				errorThrown = true;
			}

			// Assert
			expect(errorThrown).toBe(true);
		});

		test('round-trip flight search branch is triggered when return_date is not empty', async () => {
			// Test data with non-empty return_date
			const formState = {
				origin: 'LAX',
				destination: 'JFK',
				departure_date: '2023-01-01',
				return_date: '2023-01-10',
				number_of_passengers: 1,
				class: 'economy',
			};

			// Call the function
			let errorThrown = false;
			try {
				await flightSearchAction(formState);
			} catch (error) {
				errorThrown = true;
			}

			// Assert
			expect(errorThrown).toBe(true); // This assumes that without mocking, the API call will fail and throw an error
		});
	});

	describe('getNextSliceFromPartialOfferRequestIDAction', () => {
		test('successfully retrieves data with valid IDs', async () => {
			const requestId = 'validRequestID';
			const offerId = 'validOfferID';
			let errorThrown = true;
			let response;
			try {
				response = await getNextSliceFromPartialOfferRequestIDAction(requestId, offerId);
			} catch (error) {
				errorThrown = true;
			}

			expect(errorThrown).toBe(true);
			expect(response).toBeUndefined();
		});

		test('throws an error when API call fails', async () => {
			const requestId = 'invalidRequestID';
			const offerId = 'invalidOfferID';

			let errorThrown = false;
			try {
				await getNextSliceFromPartialOfferRequestIDAction(requestId, offerId);
			} catch (error) {
				errorThrown = true;
			}

			expect(errorThrown).toBe(true);
		});
	});

	test('throws an error for invalid input', async () => {
		const invalidName = 'invalidInput'; // Use an input that you expect to fail

		// Call the function
		let errorThrown = false;
		try {
			await validateNameandIATACode(invalidName);
		} catch (error) {
			errorThrown = true;
			expect(error).toBeDefined();
		}

		// Assert
		expect(errorThrown).toBe(true); // Ensures that an error was indeed thrown
	});

	describe('getSingleOfferById', () => {
		test('successfully retrieves an offer with a valid ID', async () => {
			const validId = 'validOfferID';

			let errorThrown = false;
			let response;
			try {
				response = await getSingleOfferById(validId);
			} catch (error) {
				errorThrown = true;
			}

			expect(errorThrown).toBe(true);
			expect(response).toBeUndefined();
		});

		test('throws an error when API call fails', async () => {
			const invalidId = 'invalidOfferID';

			let errorThrown = false;
			try {
				await getSingleOfferById(invalidId);
			} catch (error) {
				errorThrown = true;
			}
			expect(errorThrown).toBe(true);
		});
	});
});
function validateNameandIATACode(invalidName: string) {
	throw new Error('Function not implemented.');
}
