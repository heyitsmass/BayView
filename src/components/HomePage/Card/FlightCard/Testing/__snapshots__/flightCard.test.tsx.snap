// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FlightCard > FlightCard component snapshot 1`] = `
<div>
  <div
    class="undefined _defaultCard_c57b5c _extraCardPadding_c57b5c false
			}"
  >
    <button
      class="flex items-center -mx-3 px-1 pt-2 text-left w-full"
      data-testid="toggle-button"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-angle-down fa-lg _icon_c57b5c w-4 pr-2 icon"
        data-icon="angle-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M201.4 342.6c12.5 12.5 32.8 12.5 45.3 0l160-160c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L224 274.7 86.6 137.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l160 160z"
          fill="currentColor"
        />
      </svg>
      <div
        class="max-w-[98%]"
        style="opacity: 0; filter: blur(5px); transform: none;"
      >
        <h1
          class="text-2xl font-bold text-ellipsis whitespace-nowrap overflow-hidden"
        >
          Test
        </h1>
      </div>
    </button>
    <div
      class="box-border"
      data-testid="card-children"
      style="opacity: 0; max-height: 0px; overflow: hidden; filter: blur(10px);"
    />
  </div>
</div>
`;
